### PR TITLE
Compilation-Time Cryptocurrency selection

### PR DESCRIPTION
**Pull request by fpelliccioni**

### CODE REVIEW

File: build.py [hunk 1]
```diff
@@ -24,7 +25,17 @@
                 options["bitprim-core:with_tests"] = "True"
                 options["bitprim-core:with_examples"] = "True"
 
-            filtered_builds.append([settings, options, env_vars, build_requires])
+            opts_bch = copy.deepcopy(options)
```
Comment (dario-ramos) [inline build.py]: Why are opts_ltc commented?

File: conanfile.py [hunk 2]
```diff
@@ -127,7 +132,13 @@ def build(self):
         cmake.definitions["WITH_TESTS"] = option_on_off(self.options.with_tests)
         cmake.definitions["WITH_EXAMPLES"] = option_on_off(self.options.with_examples)
 
-        cmake.definitions["WITH_LITECOIN"] = option_on_off(self.options.with_litecoin)
+        # cmake.definitions["WITH_LITECOIN"] = option_on_off(self.options.with_litecoin)
+
+        # cmake.definitions["WITH_LTC"] = option_on_off(self.options.currency == 'LTC')
+        # cmake.definitions["WITH_BTC"] = option_on_off(self.options.currency == 'LTC')
+        # cmake.definitions["WITH_LITECOIN"] = option_on_off(self.options.currency == 'LTC')
```
Comment (dario-ramos) [inline conanfile.py:158]: Remove commented out code

Comment (dario-ramos) [reply conanfile.py:158]: (Favor using Git history to view changes)

File: include/bitcoin/bitcoin/chain/chain_state.hpp [hunk 3]
```diff
@@ -207,14 +210,16 @@ class BC_API chain_state
     // easy blocks
 
     //TODO(bitprim):
-    
+
+#ifdef BITPRIM_CURRENCY_BCH
     static uint32_t cash_difficulty_adjustment(data const& values);
+    static uint32_t work_required_adjust_cash(const data& values);
+#endif //BITPRIM_CURRENCY_BCH
```
Comment (dario-ramos) [inline include/bitcoin/bitcoin/chain/chain_state.hpp:217]: Wouldn't it be cleaner if all the cash stuff were in a separate file? (not necessarily another class, just another file)

**Review by mdallago — APPROVED** [review]

File: src/chain/chain_state.cpp [hunk 4]
```diff
@@ -569,13 +594,20 @@ uint32_t chain_state::easy_work_required(const data& values, bool daa_active)
     auto height = values.height;
     auto& bits = values.bits.ordered;
 
-    if(daa_active) 
+    if (daa_active) { 
+#ifdef BITPRIM_CURRENCY_BCH
         return cash_difficulty_adjustment(values);
-    else
+#else
+//Note: Could not happend: DAA and not BCH
```
Comment (dario-ramos) [inline src/chain/chain_state.cpp:601]: This is quite a big source file; in my opinion, this is where moving the currency-specific logic to a separate file would reduce complexity and improve readability and navigability

File: test_new/main.cpp [hunk 5]
```diff
@@ -47,4 +47,6 @@ TEST_CASE("[payment_address__construct__payment__valid_expected] payment_address
         CHECK(from_legacy.encoded_cashaddr() == from_cashaddr.encoded_cashaddr());
     }
 }
+#endif //BITPRIM_CURRENCY_BCH
```
Comment (dario-ramos) [inline test_new/main.cpp:50]: I would also place the cash-specific tests in a separate test suite

**Review by dario-ramos — APPROVED** [review]

**Review by mdallago — APPROVED** [review]

**Review by mdallago — APPROVED** [review]

**Review by mdallago — APPROVED** [review]

**Review by mdallago — APPROVED** [review]

File: include/bitcoin/bitcoin/constants.hpp [hunk 6]
```diff
@@ -91,50 +92,59 @@ BC_CONSTEXPR size_t chain_state_timestamp_count = new_bitcoin_cash_retarget_algo
...
 BC_CONSTEXPR size_t max_sigops_factor = 50;
-BC_CONSTEXPR size_t max_block_sigops = max_block_size / max_sigops_factor; 
-BC_CONSTEXPR size_t max_block_sigops_cash = max_block_size_cash / max_sigops_factor;
 
+#ifdef BITPRIM_CURRENCY_BCH
+// BC_CONSTEXPR size_t max_block_sigops_cash = max_block_size_cash / max_sigops_factor;
+BC_CONSTEXPR size_t max_block_sigops = max_block_size / max_sigops_factor;
+#else
+BC_CONSTEXPR size_t max_block_sigops = max_block_size / max_sigops_factor; 
+#endif //BITPRIM_CURRENCY_BCH
```
Comment (ramacarlucho) [inline include/bitcoin/bitcoin/constants.hpp]: Both expr are the same

**Review by ramacarlucho — APPROVED** [review]

File: CMakeLists.txt [hunk 7]
```diff
@@ -557,7 +582,8 @@ target_compile_definitions(bitprim-core PUBLIC -DBITPRIM_BUILD_NUMBER="\\"${BITP
 
 target_include_directories(bitprim-core SYSTEM PUBLIC ${Boost_INCLUDE_DIR})
 
-if (WITH_LITECOIN)
+# if (WITH_LITECOIN)
```
Comment (dario-ramos) [inline CMakeLists.txt:579]: Duplicates lines 568-569

**Review by dario-ramos — APPROVED** [review]

File: include/bitcoin/bitcoin/constants.hpp (hunk 6 above)
Comment (fpelliccioni) [reply include/bitcoin/bitcoin/constants.hpp]: Thanks, I will merge them.

File: CMakeLists.txt (hunk 7 above)
Comment (fpelliccioni) [reply CMakeLists.txt:579]: Solved. Thanks!

File: include/bitcoin/bitcoin/constants.hpp (hunk 6 above)
Comment (fpelliccioni) [reply include/bitcoin/bitcoin/constants.hpp]: Solved! Thanks!

**Review by mdallago — APPROVED** [review]

**Pull request merged**